videostore/indexer.go

```diff
@@ -0,0 +1,392 @@
+package videostore
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	// Blank importing SQLite3 driver for its side-effects.
+	// This registers the "sqlite3" driver with the database/sql package.
+	_ "github.com/mattn/go-sqlite3"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/utils"
+)
+
+const (
+	dbFileName        = "index.sqlite.db"
+	pollingInterval   = 5 * time.Second
+	segmentsTableName = "segments"
+	dbFileMode        = 0o750
+	videoFileSuffix   = ".mp4"
+)
+
+// indexer manages metadata for video segments stored on disk.
+type indexer struct {
+	logger      logging.Logger
+	storagePath string
+	dbPath      string
+	db          *sql.DB
+	setupDone   bool
+}
+
+// segmentMetadata holds metadata for an indexed segment.
+type segmentMetadata struct {
+	FilePath  string
+	SizeBytes int64
+}
+
+// newIndexer creates a new indexer instance.
+func newIndexer(storagePath string, logger logging.Logger) *indexer {
+	dbPath := filepath.Join(storagePath, dbFileName)
+	return &indexer{
+		logger:      logger,
+		storagePath: storagePath,
+		dbPath:      dbPath,
+	}
+}
+
+func (ix *indexer) setup() error {
+	if ix.setupDone {
+		return nil
+	}
+	if err := os.MkdirAll(filepath.Dir(ix.dbPath), dbFileMode); err != nil {
+		return fmt.Errorf("failed to create directory for index db: %w", err)
+	}
+	db, err := sql.Open("sqlite3", ix.dbPath)
+	if err != nil {
+		return fmt.Errorf("failed to open index db: %w", err)
+	}
+	ix.db = db
+	if err := ix.initializeDB(); err != nil {
+		_ = ix.db.Close()
+		ix.db = nil
+		return fmt.Errorf("failed to initialize index db schema: %w", err)
+	}
+	ix.setupDone = true
+	return nil
+}
+
+func (ix *indexer) initializeDB() error {
+	query := fmt.Sprintf(`
+	CREATE TABLE IF NOT EXISTS %s (
+		file_path TEXT PRIMARY KEY,
+		start_time_unix INTEGER NOT NULL,
+		duration_ms INTEGER,
```

Comment:
Could we consider preemptively adding rows for the rest of `videoInfo`(width, height, and codec)

```type videoInfo struct {
	duration time.Duration
	width    int
	height   int
	codec    string
}
```

The utility [matchStorageToRange](https://github.com/viam-modules/video-store/blob/main/videostore/utils.go#L269) could be refactored a bit to use the indexer instead.

I think this could be a follow-up ticket, but not sure about the complications of migrating schemas once things go live.

Reply:
I'm down if it makes things easier down the line

Reply:
@nicksanford What do you think about adding additional videoInfo attrs to the database?